### PR TITLE
Lexer. Skip 4 whitespaces at a time

### PR DESCRIPTION
**Pull request by welkam**
Made lexer skip 4 spaces at a time after newline. Here are some benchmarks lexing DMD source code.
Current lexer:
dmd                213149 usecs
dmd -O -inline 137161 usecs
ldc --O3           108218 usecs

New improved:
dmd                194221 usecs    9%  faster
dmd -O -inline 123456 usecs  10% faster
ldc --O3            90630 usecs   16% faster

Overall compilation speed might speed up by 1% for simple code.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]
Nice!

File: src/dmd/lexer.d [hunk 1]
```diff
@@ -373,11 +373,19 @@ class Lexer
...
+                {
                     endOfLine();
+                    goto skipFourSpaces;
+                }
                 continue; // skip white space
             case '\n':
                 p++;
                 endOfLine();
+                skipFourSpaces:
+                while (*(cast(uint*)p) == 0x20202020) //' ' == 0x20
```
Comment (wilzbach) [inline src/dmd/lexer.d:385]: Shouldn't this depend on the architecture?
i.e. on 64-bit you could compare with 8 byte (or here whitespace) at a time?

Comment (welkam) [reply src/dmd/lexer.d:385]: Most editors when you press tab write 4 spaces unless you change the settings. If I checked 8 bytes at a time I would miss optimization for 1 tab(common) and reduced optimization for odd number of tabs. While its possible to make use of bigger registers on 64 bit machines I dont believe increased code complexity is worth it.

**Comment by kinke** [comment]
AFAICT, this might read beyond the file buffer; by the looks of it, only 2 null bytes are appended to the file contents: https://github.com/dlang/dmd/blob/43729283fe0149b171e4f3a2c60f8cf5f78b701d/src/dmd/root/file.d#L120-L121

So appending 2 more should probably suffice.

**Comment by welkam** [comment]
In rare instance that allocated buffer ends at page boundary a read past the buffer can cause segmentation fault. Because of that we need to increase buffer by 2 more bytes.

File: src/dmd/lexer.d (hunk 1 above)
Comment (dkorpel) [reply src/dmd/lexer.d:385]: `p` is a byte-aligned pointer right?
Is it defined behavior to dereference a `uint*` that is not 4-byte aligned?

Comment (kinke) [reply src/dmd/lexer.d:385]: Ah yeah that's another potential issue, at least for non-x86 archs.

Comment (welkam) [reply src/dmd/lexer.d:385]: From GNU compiler
"Enables (or disables) reading and writing of 16- and 32- bit values from addresses that are not 16- or 32- bit aligned. By default unaligned access is disabled for all pre-ARMv6, all ARMv6-M and for ARMv8-M Baseline architectures, and enabled for all other architectures"

ARMv7 and ARMv8 processors that are in phones and tablets should work.
https://gcc.gnu.org/onlinedocs/gcc/ARM-Options.html

**Comment by ibuclaw** [comment]
This PR broke the SPARC compiler.

**Comment by ibuclaw** [comment]
https://issues.dlang.org/show_bug.cgi?id=22517

**Pull request merged**